rapids-cmake/test/generate_resource_spec.cmake

```diff
@@ -60,7 +60,8 @@ function(rapids_test_generate_resource_spec DESTINATION filepath)
 
     set(compiler "${CMAKE_CXX_COMPILER}")
     set(lang CXX)
-    if(DEFINED CMAKE_CUDA_COMPILER)
+    get_property(rapids_languages GLOBAL PROPERTY ENABLED_LANGUAGES)
+    if("CUDA" IN_LIST rapids_languages)
```

Comment:
Reading over the logs of the original bug report, I'm not sure this will fully fix it. In particular, the `detail/generate_resource_spec/CMakeLists.txt` has these lines:
```cmake
if(cuda_toolkit)
  find_package(CUDAToolkit REQUIRED QUIET) # <-- failed here
  target_link_libraries(generate_ctest_json PRIVATE CUDA::cudart_static)
  # ...
```
So the problem is that `cuda_toolkit` was truthy (as the first `find_package()` here in `generate_resource_spec.cmake` found... something) even though CUDA support was disabled.

So if `CUDA` is not in the enabled languages, consider also
```cmake
set(CUDAToolkit_FOUND OFF)
```

Reply:
CUDA language support is orthogonal to CUDAToolkit support. You can have a C++ compiler only project that uses the CUDA Toolkit but doesn't launch CUDA kernels directly.

So we aren't going to enforce requiring the CUDA language to be enabled to  use the CUDAToolkit or to use `generate_resource_spec`.  It shouldn't be a problem is CUDAToolkit is found with C++ as we have been doing here since the start

Reply:
> So we aren't going to enforce requiring the CUDA language to be enabled to use the CUDAToolkit or to use generate_resource_spec. It shouldn't be a problem is CUDAToolkit is found with C++ as we have been doing here since the start

Yeah but that is precisely the bug. I have tracked the erroneous logic down as follows:

1. `CUDA` is explicitly disabled by the user, i.e. `get_propert(ENABLED_LANGUAGES)` will not contain "CUDA".
1. `generate_resource_spec.cmake` does `find_package(CUDAToolkit QUIET)`. This _succeeds_.
2. `generate_resource_spec.cmake` thus has `CUDAToolkit_FOUND` set to a truthy value.
3. `generate_resource_spec.cmake` passes this to `detail/generate_resource_spec/CMakeLists.txt` via `-Dcuda_toolkit=${CUDAToolkit_FOUND}`.
4. `detail/generate_resource_spec/CMakeLists.txt` then sees `cuda_toolkit=TRUE`, and does `find_package(CUDAToolkit REQUIRED). This _fails_.

In this case, CUDAToolkit was able to be found in the "main" cmake lists (probably due to some extra variables set by `conda`), but subsequently not found in the resource generation `CMakeLists.txt`.

So the issue is, we are using the wrong canary to determine whether CUDA support is both enabled _and_ requested.

Reply:
> So the issue is, we are using the wrong canary to determine whether CUDA support is both enabled and requested.

The issue is that we aren't passing sufficient information to the try_compile to support `find_package(CUDAToolkit)`. I know this is pendantic but we need to use the correct terms. The root project correctly found the CUDA Toolkit without the CUDA language enabled. 

When using previous versions of CMake what was the end result of the `generate_resource_spec`? Did it generate a json file with `HAVE_CUDA` being defined or not? I am thinking that the problematic part is:

```
generate_resource_spec.cmake does find_package(CUDAToolkit QUIET). This succeeds.
```
With previous CMake versions this would fail since the `CMAKE_CUDA_COMPILER` ( and ENV{CUDACXX} ) had not been used to help find the CUDAToolkit. Now we do find the CUDAToolkit since that information was provided to us, but the inner `try_compile` is failing as we don't propagate `CMAKE_CUDA_COMPILER` down.

Reply:
Hmmm, so perhaps we should propagate `CMAKE_CUDA_COMPILER` down unconditionally then? I suppose it cannot hurt:

1. If `CUDAToolkit_FOUND` is falsey, nothing will inspect it
2. If an older CMake version is used, then it won't consult the variable in the search.
3. If the newest CMake is used, it will help the `find_package(CTK)` to succeed in both places.

Reply:
That is my thinking as well. The `CUDACXX` env variable should already be auto propagated since the `try_compile` is a sub-process of CMake itself...